### PR TITLE
robust solver updates

### PR DESCRIPTION
**Pull request by ulricharmel**
I have decided to push all the small changes I have made to the robust solver recently.

### CODE REVIEW

**Comment by JSKenyon** [comment]
@ulricharmel Please merge the current CubiCal master into your branch. This currently includes months of commits that have already gone into master.

**Comment by ulricharmel** [comment]
@JSKenyon ok let me do that

File: cubical/data_handler/ms_tile.py [hunk 1]
```diff
@@ -644,20 +644,22 @@ def fill_bitflags(self, flagbit):
             flagcol = table_subset.getcol("FLAG")
             flagrow = table_subset.getcol("FLAG_ROW")
             # flag all four corrs
+
```
Comment (JSKenyon) [inline cubical/data_handler/ms_tile.py:647]: Please minimise the diff by removing the added whitespacing here.

File: cubical/machines/interval_gain_machine.py [hunk 2]
```diff
@@ -196,10 +196,12 @@ def determine_allocators(cls, options):
 
     @classmethod
     def get_full_kernel(cls, options, diag_gains):
+
```
Comment (JSKenyon) [inline cubical/machines/interval_gain_machine.py:199]: Again, please minimise the diff by removing added whitespace and testing artifacts.

File: cubical/machines/jones_chain_machine.py [hunk 3]
```diff
@@ -487,6 +487,7 @@ def next_iteration(self):
 
     def compute_chisq(self, resid_arr, inv_var_chan, require_full=True):
         """Computes chi-square using the active chain term"""
+
```
Comment (JSKenyon) [inline cubical/machines/jones_chain_machine.py:490]: Remove whitespace.

File: cubical/madmax/plots.py [hunk 4]
```diff
@@ -17,9 +17,11 @@ def make_dual_absres_plot(absres, fl_prior, fl_new, p, q, metadata, subplot_titl
     resmask = np.zeros_like(absres[0, :, :, p, q], dtype=bool)
     resmask[:] = fl_prior[..., np.newaxis, np.newaxis] #| absres[0, :, :, p, q]==0
     res = np.ma.masked_array(absres[0, :, :, p, q], resmask)
+
```
Comment (JSKenyon) [inline cubical/madmax/plots.py:20]: Remove unnecessary whitespace please.

File: cubical/main.py [hunk 5]
```diff
@@ -536,7 +536,7 @@ def prelog_print(level, message):
         chunks_per_tile, tile_list = ms.define_chunk(GD["data"]["time-chunk"], GD["data"]["rebin-time"],
                                             GD["data"]["freq-chunk"],
                                             chunk_by=chunk_by, chunk_by_jump=jump,
-                                            chunks_per_tile=chunks_per_tile, max_chunks_per_tile=max_chunks_per_tile)
+                                            chunks_per_tile=chunks_per_tile, max_chunks_per_tile=GD["dist"]["max-chunks"])
```
Comment (JSKenyon) [inline cubical/main.py:539]: What is this change? Looks suspicious to me.

File: cubical/main.py [hunk 6]
```diff
@@ -669,3 +669,5 @@ def prelog_print(level, message):
                 pdb.post_mortem(tb)
         sys.exit(2 if type(exc) is UserInputError else 1)
 
+
```
Comment (JSKenyon) [inline cubical/main.py:672]: Get rid of whitespace.

File: cubical/solver.py [hunk 7]
```diff
@@ -85,6 +85,9 @@ def _solve_gains(gm, stats, madmax, obser_arr, model_arr, flags_arr, sol_opts, l
 
     diverging = ""
 
+    # for cases the are no valid solutions
+    gm.output_weights = None
```
Comment (JSKenyon) [inline cubical/solver.py:89]: Is this safe?

File: cubical/solver.py [hunk 8]
```diff
@@ -400,11 +409,20 @@ def compute_chisq(statfield=None, full=True):
         resid_arr = obser_arr
 
     robust_weights = None
-    if hasattr(gm, 'save_weights'):
+    if hasattr(gm, 'is_robust'):
+        
+        # do a last round of robust flag robust flag and save the weights
+         
+        if gm.robust_flag_weights and not gm.flag_disable_by_sover:
```
Comment (JSKenyon) [inline cubical/solver.py:416]: Please change variable name - gm.flags_disabled_by_solver

File: cubical/solver.py [hunk 9]
```diff
@@ -710,6 +727,8 @@ def run(self):
                      self.vdm.weighted_obser, self.vdm.weighted_model, self.vdm.flags_arr,
                      self.sol_opts, label=self.label)
 
+        self.output_weights = self.gm.output_weights
```
Comment (JSKenyon) [inline cubical/solver.py:730]: Is this safe? I have no intuition for this.

File: cubical/solver.py (hunk 7 above)
Comment (ulricharmel) [reply cubical/solver.py:89]: Yes this and the one below safe. i.e. self.output_weights = self.gm.output_weights
Actually initializing the weights to None is for the complex solver and all the other solvers which do not output weights.

File: cubical/solver.py (hunk 9 above)
Comment (ulricharmel) [reply cubical/solver.py:730]: It is safe, I added this because the return statement in _solver_gains was change. So this is the only way return and save the weights from the robust solver in the MS

File: cubical/main.py (hunk 5 above)
Comment (ulricharmel) [reply cubical/main.py]: I have updated it to this to the one on master.

**Comment by JSKenyon** [comment]
ok to test

**Comment by JSKenyon** [comment]
@ulricharmel This has been merged. Please be sure to delete the branch on your fork and sync up with master. This will just avoid 100+ commit PRs.

**Comment by ulricharmel** [comment]
@JSKenyon  sure I will do that now.
